### PR TITLE
sensors: add lm77 digital temperature sensor driver

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Add driver for the TI LM77 digital temperature sensor with two-wire interface.

The driver supports the following features:
- Ambient temperature sensor channel.
- Threshold trigger (T_low, T_high, and T_hyst).
- External critical temperature alert (T_crit).
- Power management (shutdown).
- Full configuration register support via DTS.

### CODE REVIEW

**Review by JordanYates — CHANGES_REQUESTED** [review]

File: drivers/sensor/lm77/lm77.c [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+		uint8_t int_mode : 1;
+		uint8_t tcrita_pol : 1;
+		uint8_t int_pol : 1;
+		uint8_t fault_queue : 1;
+		uint8_t reserved : 3;
+	} __packed;
+};
+
+struct lm77_config {
+	const struct device *i2c;
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply drivers/sensor/lm77/lm77.c]: Updated, thanks.

File: dts/bindings/sensor/lm77.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) 2021 Vestas Wind Systems A/S
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  LM77 digital temperature sensor
+
+compatible: "lm77"
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply dts/bindings/sensor/lm77.yaml:7]: The `compatible` is kept without vendor prefix to remain compatible with Linux devicetree binding (https://elixir.bootlin.com/linux/v5.13.8/source/drivers/hwmon/lm77.c#L349).

File: drivers/sensor/lm77/lm77.c [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+
+LOG_MODULE_REGISTER(lm77, CONFIG_SENSOR_LOG_LEVEL);
+
+/*
+ * Only compile in trigger support if enabled in Kconfig and at least one
+ * enabled lm77 devicetree node has the int-gpios property.
+ */
+#define LM77_TRIGGER_SUPPORT_INST(inst) DT_INST_NODE_HAS_PROP(inst, int_gpios) ||
+#define LM77_TRIGGER_SUPPORT \
+	(CONFIG_LM77_TRIGGER && (DT_INST_FOREACH_STATUS_OKAY(LM77_TRIGGER_SUPPORT_INST) 0))
```
Comment (gmarull) [inline drivers/sensor/lm77/lm77.c:24]: nice trick, wasn't aware of it

File: drivers/sensor/lm77/lm77.c [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+	struct lm77_data *data = dev->data;
+	int16_t temp;
+	int err;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_AMBIENT_TEMP) {
+		return -ENOTSUP;
+	}
+
+	err = lm77_read_temp(dev, LM77_REG_TEMP, &temp);
+	if (err) {
```
Comment (gmarull) [inline drivers/sensor/lm77/lm77.c:276]: nit: better use an explicit boolean expression, e.g. `err < 0` (ref. MISRA 14.4)

**Review by gmarull — APPROVED** [review]
LGTM, left one minor comment

File: drivers/sensor/lm77/lm77.c (hunk 4 above)
Comment (henrikbrixandersen) [reply drivers/sensor/lm77/lm77.c]: @gmarull Thanks. I have changed all `if` expressions to be _essentially boolean_ type.

File: drivers/sensor/lm77/lm77.c (hunk 3 above)
Comment (gmarull) [reply drivers/sensor/lm77/lm77.c:24]: I think VARGS macro can also be used to introduce the `||` operation:
```suggestion
	(CONFIG_LM77_TRIGGER && (DT_INST_FOREACH_STATUS_OKAY_VARGS(LM77_TRIGGER_SUPPORT_INST, ||) 0))
```

https://github.com/zephyrproject-rtos/zephyr/pull/37483 adds some example docs for the non-inst version

File: drivers/sensor/lm77/lm77.c [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+
+static int lm77_read_temp(const struct device *dev, uint8_t reg, int16_t *value)
+{
+	const struct lm77_config *config = dev->config;
+	uint8_t buf[2];
+	int err;
+
+	err = i2c_write_read_dt(&config->i2c, &reg, sizeof(reg), &buf, sizeof(buf));
+	if (err < 0) {
+		return err;
```
Comment (MaureenHelm) [inline drivers/sensor/lm77/lm77.c:85]: Suggest adding a LOG_ERR here

**Review by MaureenHelm — APPROVED** [review]

File: drivers/sensor/lm77/lm77.c (hunk 5 above)
Comment (henrikbrixandersen) [reply drivers/sensor/lm77/lm77.c:86]: @MaureenHelm Added, thanks.

**Comment by henrikbrixandersen** [comment]
@JordanYates Please re-evaluate.

**Review by JordanYates — APPROVED** [review]

**Pull request merged**